### PR TITLE
(PUP-9717) Retry state machine on errors

### PR DESCRIPTION
**Pull request by joshcooper**
Backports several commits from master: 81e2668, 5c7bfb7, c487c2f 

Updates the state machine to retry when exceptions occur, not just when the CSR hasn't been signed yet. By default the state machine will either reach the Done state or keep retrying, waiting every `waitforcert` seconds in between.

Preserves the behavior where if we ever try to wait, but `waitforcert < 1`, then exit.

Adds a `onetime` option to the state machine so it will raise if an error occurs, and never retry. This is needed for things like `puppet agent --test/--onetime`, which don't retry on errors. It is also needed to preserve the legacy `Puppet::SSL::Host#certificate` behavior.

### CODE REVIEW

**Comment by joshcooper** [comment]
Some examples of scenarios I've tested:

```
$ bx puppet agent --no-daemonize --ca_port 666 --logdest console --waitforcert 2
Error: Failed to open TCP connection to puppet.delivery.puppetlabs.net:666 (Connection refused - connect(2) for "puppet.delivery.puppetlabs.net" port 666)
Error: Connection to cached server and port puppet.delivery.puppetlabs.net:666 failed: Failed to open TCP connection to puppet.delivery.puppetlabs.net:666 (Connection refused - connect(2) for "puppet.delivery.puppetlabs.net" port 666)
Error: Connection to cached server and port puppet.delivery.puppetlabs.net:666 failed: Failed to open TCP connection to puppet.delivery.puppetlabs.net:666 (Connection refused - connect(2) for "puppet.delivery.puppetlabs.net" port 666)
...
```

```
$ bx puppet agent --test --ca_port 666
Info: csr_attributes file loading from /Users/josh/.puppetlabs/etc/puppet/csr_attributes.yaml
Info: Creating a new SSL certificate request for neptune
Info: Certificate Request fingerprint (SHA256): 44:95:77:61:F4:51:C4:35:4A:28:20:47:1E:4F:5F:68:7D:8A:FA:58:FA:1B:4B:F5:A9:33:4A:8F:D7:34:37:9D
Error: Failed to open TCP connection to puppet.delivery.puppetlabs.net:666 (Connection refused - connect(2) for "puppet.delivery.puppetlabs.net" port 666)
Error: Could not run: Failed to open TCP connection to puppet.delivery.puppetlabs.net:666 (Connection refused - connect(2) for "puppet.delivery.puppetlabs.net" port 666)
$
```

```
$ bx puppet agent --no-daemonize --onetime --ca_port 666 --logdest console
Error: Failed to open TCP connection to puppet.delivery.puppetlabs.net:666 (Connection refused - connect(2) for "puppet.delivery.puppetlabs.net" port 666)
Error: Could not run: Failed to open TCP connection to puppet.delivery.puppetlabs.net:666 (Connection refused - connect(2) for "puppet.delivery.puppetlabs.net" port 666)
$
```

```
$ bx puppet ssl bootstrap  --ca_port 666
Info: csr_attributes file loading from /Users/josh/.puppetlabs/etc/puppet/csr_attributes.yaml
Info: Creating a new SSL certificate request for neptune
Info: Certificate Request fingerprint (SHA256): 44:95:77:61:F4:51:C4:35:4A:28:20:47:1E:4F:5F:68:7D:8A:FA:58:FA:1B:4B:F5:A9:33:4A:8F:D7:34:37:9D
Error: Failed to open TCP connection to puppet.delivery.puppetlabs.net:666 (Connection refused - connect(2) for "puppet.delivery.puppetlabs.net" port 666)
Info: Couldn't fetch certificate from CA server; you might still need to sign this agent's certificate (neptune). Will try again in 120 seconds.
Error: Failed to open TCP connection to puppet.delivery.puppetlabs.net:666 (Connection refused - connect(2) for "puppet.delivery.puppetlabs.net" port 666)
Info: Couldn't fetch certificate from CA server; you might still need to sign this agent's certificate (neptune). Will try again in 120 seconds.
...
```

Note `puppet agent --fingerprint` is still not right in the case where the agent doesn't have a client cert yet, and that will be fixed in PUP-9720.

**Comment by joshcooper** [comment]
jenkins please test this with server tests on windows2016-64a

**Comment by joshcooper** [comment]
After working on the fingerprint issue, I don't think commit https://github.com/puppetlabs/puppet/pull/7560/commits/e611f510a2c273224ca3e5001b58c74e013b05e7 is necessary and can be removed. I'll update that.

File: spec/unit/ssl/state_machine_spec.rb [hunk 1]
```diff
@@ -61,6 +102,69 @@
...
+        stub_request(:get, %r{puppet-ca/v1/certificate/#{Puppet[:certname]}})
+          .to_return(status: 404).then
+          .to_return(status: 200, body: client_cert.to_pem)
+        stub_request(:put, %r{puppet-ca/v1/certificate_request/#{Puppet[:certname]}})
+          .to_raise(Errno::ECONNREFUSED).then
+          .to_return(status: 200)
+
+        machine.ensure_client_certificate
+
+        expect(@logs).to include(an_object_having_attributes(message: refused_message))
```
Comment (kris-bosland) [inline spec/unit/ssl/state_machine_spec.rb]: Not clear how this shows retries...?  I can see from the later test that you can show here you are not getting an error.  Maybe make sure it got the client_cert?

File: lib/puppet/ssl/state_machine.rb [hunk 2]
```diff
@@ -17,8 +17,14 @@ class SSLState
     def initialize(machine, ssl_context)
       @machine = machine
       @ssl_context = ssl_context
-      @cert_provider = Puppet::X509::CertProvider.new
-      @ssl_provider = Puppet::SSL::SSLProvider.new
+      @cert_provider = machine.cert_provider
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply lib/puppet/ssl/state_machine.rb]: yeah it's the passed in machine, not `self`

File: spec/unit/ssl/state_machine_spec.rb (hunk 1 above)
Comment (joshcooper) [reply spec/unit/ssl/state_machine_spec.rb]: I added a commit to verify that the requests are retried

**Pull request merged**